File: spec/probe-types.ts
/**
 * UI Test Probe — Authoritative TypeScript Type Definitions
 *
 * This file is the single source of truth for all probe types.
 * All platform SDKs and integrations import from here.
 */

// ---------------------------------------------------------------------------
// ProbeType — 9 widget categories
// ---------------------------------------------------------------------------

export enum ProbeType {
  DATA_CONTAINER = 'data-container',
  SELECTOR = 'selector',
  ACTION = 'action',
  DISPLAY = 'display',
  MEDIA = 'media',
  FORM = 'form',
  PAGE = 'page',
  MODAL = 'modal',
  NAVIGATION = 'navigation',
}

// ---------------------------------------------------------------------------
// Linkage
// ---------------------------------------------------------------------------

export enum LinkageEffect {
  OPTIONS_UPDATE = 'options_update',
  DATA_RELOAD = 'data_reload',
  VISIBILITY_TOGGLE = 'visibility_toggle',
  ENABLED_TOGGLE = 'enabled_toggle',
  VALUE_UPDATE = 'value_update',
  RESET = 'reset',
  NAVIGATE = 'navigate',
}

export type LinkagePath =
  | { type: 'direct' }
  | { type: 'api'; url: string; method?: string }
  | { type: 'computed'; expression: string }
  | { type: 'store'; storeName: string; action?: string }
  | { type: 'navigation'; route: string }
  | { type: 'chain'; through: string };

export interface LinkageTarget {
  id: string;
  effect: LinkageEffect;
  path: LinkagePath;
}

export interface LinkageResult {
  trigger: string;
  action: string;
  directEffects: Array<{
    target: string;
    effect: string;
    result: 'pass' | 'fail' | 'timeout';
    duration: number;
  }>;
  chainedEffects: Array<{
    target: string;
    effect: string;
    through: string;
    result: 'pass' | 'fail' | 'timeout';
    duration: number;
  }>;
  apiCalls: Array<{
    url: string;
    method: string;
    status: number;
    responseTime: number;
  }>;
}

// ---------------------------------------------------------------------------
// ProbeElement — 15 attributes
// ---------------------------------------------------------------------------

export interface ProbeElement {
  /** Semantic identifier, e.g. "order-list". */
  id: string;

  /** Widget type — determines data field shape. */
  type: ProbeType;

  /** Accessibility metadata. */
  accessibility?: {
    role?: string;
    label?: string;
    tabIndex?: number;
  };

  /** Real-time element state. */
  state: {
    current: string;
    previous?: string;
    timestamp: number;
    isOpen?: boolean;
    validationErrors?: Array<{ field: string; message: string }>;
  };

  /**
   * Data payload — structure varies by type.
   * data-container: sort, filter, selectedRows
   * media: currentTime, duration, readyState, paused, networkState
   */
  data?: {
    value?: unknown;
    options?: string[];
    rows?: number;
    columns?: Array<{ id: string; label: string; visible: boolean }>;

    // data-container extensions
    sort?: { column: string; direction: 'asc' | 'desc' };
    filter?: Array<{ field: string; operator: string; value: unknown }>;
    selectedRows?: string[];

    // media extensions
    currentTime?: number;
    duration?: number;
    readyState?: number;
    paused?: boolean;
    networkState?: number;
  };

  /** Data source binding (API url, status, timing). */
  source?: {
    url: string;
    method: string;
    status?: number;
    responseTime?: number;
    payload?: unknown;
  };

  /** Inter-element linkage declarations. */
  linkage?: {
    targets: LinkageTarget[];
  };

  /** Layout metrics from the rendering engine. */
  layout: {
    x: number;
    y: number;
    width: number;
    height: number;
    visible: boolean;
    renderTime?: number;
    scrollTop?: number;
    scrollLeft?: number;
  };

  /** Keyboard shortcuts bound to this element. */
  shortcuts?: Array<{
    key: string;
    action: string;
    platform?: string;
  }>;

  /** Animation state. */
  animation?: {
    playing: boolean;
    name?: string;
    duration?: number;
    progress?: number;
  };

  /** Locale / directionality. */
  locale?: {
    language?: string;
    direction?: 'ltr' | 'rtl';
    isRTL?: boolean;
  };

  /** Theme context. */
  theme?: {
    mode?: 'light' | 'dark' | 'high-contrast';
    colorScheme?: string;
  };

  /** DOM event types bound to this element. */
  eventBindings?: string[];

  /** Session/form dirty state. */
  session?: {
    isDirty?: boolean;
    hasUnsavedChanges?: boolean;
  };

  /** Parent element probe ID. */
  parent?: string;

  /** Child element probe IDs. */
  children?: string[];
}

// ---------------------------------------------------------------------------
// Platform Context
// ---------------------------------------------------------------------------

export enum Platform {
  WEB_CHROME = 'web-chrome',
  WEB_SAFARI = 'web-safari',
  WEB_FIREFOX = 'web-firefox',
  IOS = 'ios',
  ANDROID = 'android',
  MACOS = 'macos',
  WINDOWS = 'windows',
  LINUX = 'linux',
}

export type FormFactor = 'phone' | 'tablet' | 'desktop' | 'foldable';

export interface DeviceProfile {
  name: string;
  screenSize: { width: number; height: number };
  pixelRatio: number;
  hasNotch: boolean;
  hasSafeArea: boolean;
  formFactor: FormFactor;
}

export enum InputMode {
  TOUCH = 'touch',
  MOUSE_KEYBOARD = 'mouse_keyboard',
  STYLUS = 'stylus',
  GAMEPAD = 'gamepad',
}

export interface PlatformContext {
  platform: Platform;
  device: DeviceProfile;
  viewport: { width: number; height: number };
  inputMode: InputMode;
}

/** Named device preset key (e.g. "iphone-15-pro"). */
export type ViewportPreset =
  | 'iphone-se'
  | 'iphone-15-pro'
  | 'ipad-air'
  | 'ipad-pro-12'
  | 'pixel-8'
  | 'galaxy-s24'
  | 'galaxy-tab-s9'
  | 'galaxy-fold'
  | 'macbook-air-13'
  | 'desktop-1080p'
  | 'desktop-1440p';

// ---------------------------------------------------------------------------
// Snapshots
// ---------------------------------------------------------------------------

export interface ProbeSnapshot {
  timestamp: number;
  elements: Record<string, ProbeElement>;
  platformContext: PlatformContext;
}

export interface SnapshotDiff {
  elementId: string;
  field: string;
  before: unknown;
  after: unknown;
}

// ---------------------------------------------------------------------------
// Network log entry
// ---------------------------------------------------------------------------

export interface NetworkLogEntry {
  url: string;
  method: string;
  status: number;
  elementId?: string;
  timestamp: number;
}

// ---------------------------------------------------------------------------
// Test result (for runAcrossDevices)
// ---------------------------------------------------------------------------

export interface TestResult {
  device: string;
  passed: boolean;
  duration: number;
  error?: string;
}

// ---------------------------------------------------------------------------
// Action pre-check error
// ---------------------------------------------------------------------------

export interface ProbeActionError {
  error: 'NOT_FOUND' | 'NOT_VISIBLE' | 'DISABLED' | 'BUSY' | 'OPTION_NOT_FOUND';
  id: string;
  reason?: string;
  available?: string[];
}

// ---------------------------------------------------------------------------
// actAndWait result
// ---------------------------------------------------------------------------

export interface ActAndWaitResult {
  actionDuration: number;
  waitDuration: number;
  targetState: ProbeElement['state'];
  linkageResults?: LinkageResult;
}

// ---------------------------------------------------------------------------
// Page query result
// ---------------------------------------------------------------------------

export interface PageQueryResult {
  id: string;
  state: string;
  elements: ProbeElement[];
  unreadyElements: string[];
}

// ---------------------------------------------------------------------------
// UITestProbe — full interface (all 6 primitives + hierarchy + platform)
// ---------------------------------------------------------------------------

export interface UITestProbe {
  // --- Primitive 1: Element Registry ---
  query(id: string): ProbeElement | null;
  queryAll(type?: ProbeType): ProbeElement[];
  queryPage(): PageQueryResult;

  // --- Primitive 1.5: Platform Context ---
  setPlatformContext(context: PlatformContext): Promise<void>;
  getPlatformContext(): PlatformContext;
  setDevice(preset: ViewportPreset | string): Promise<void>;
  runAcrossDevices(
    devices: Array<ViewportPreset | string>,
    test: (probe: UITestProbe) => Promise<void>,
  ): Promise<Record<string, TestResult>>;

  // --- Primitive 1.6: Hierarchy ---
  queryChildren(id: string): ProbeElement[];
  queryDescendants(id: string): ProbeElement[];
  queryParent(id: string): ProbeElement | null;
  getAncestorChain(id: string): ProbeElement[];
  isEffectivelyVisible(id: string): boolean;

  // --- Primitive 2: State Exposure ---
  getState(id: string): ProbeElement['state'];
  getStates(ids: string[]): Record<string, ProbeElement['state']>;

  // --- Primitive 3: Event Stream ---
  waitForState(id: string, state: string, timeout?: number): Promise<void>;
  waitForPageReady(timeout?: number): Promise<void>;
  onStateChange(id: string, callback: (oldState: string, newState: string) => void): () => void;
  onEvent(id: string, event: string, callback: (detail: unknown) => void): () => void;

  // --- Primitive 4: Source Binding ---
  getSource(id: string): ProbeElement['source'] | null;
  getNetworkLog(): NetworkLogEntry[];
  waitForSource(id: string, timeout?: number): Promise<ProbeElement['source']>;

  // --- Primitive 5: Layout Metrics ---
  getLayout(id: string): ProbeElement['layout'];
  getOverlaps(): Array<{ a: string; b: string; overlapArea: number }>;
  getScrollPosition(id: string): { scrollTop: number; scrollLeft: number };

  // --- Primitive 6: Action Dispatch ---

  // Basic interactions
  click(id: string): Promise<void>;
  doubleClick(id: string): Promise<void>;
  rightClick(id: string): Promise<void>;
  hover(id: string): Promise<void>;
  focus(id: string): Promise<void>;

  // Input
  type(id: string, text: string): Promise<void>;
  fill(id: string, value: string): Promise<void>;
  clear(id: string): Promise<void>;

  // Selection
  select(id: string, value: string): Promise<void>;
  check(id: string, checked: boolean): Promise<void>;

  // Scroll
  scrollTo(id: string, position: { top?: number; left?: number }): Promise<void>;
  scrollToBottom(id: string): Promise<void>;
  scrollIntoView(id: string): Promise<void>;

  // Drag & drop
  drag(sourceId: string, targetId: string): Promise<void>;

  // Keyboard
  pressShortcut(key: string): Promise<void>;

  // Navigation
  navigate(route: string): Promise<void>;

  // Composite
  snapshot(): ProbeSnapshot;
  diff(a: ProbeSnapshot, b: ProbeSnapshot): SnapshotDiff[];
  verifyLinkage(triggerId: string, action: string): Promise<LinkageResult>;
  actAndWait(
    id: string,
    action: string,
    waitFor: { target?: string; state?: string; timeout?: number },
  ): Promise<ActAndWaitResult>;
}

File: sdk/windows/Models/ProbeTypes.cs
namespace UITestProbe.Models;

/// <summary>
/// Control type classification. Maps to the unified ProbeType enum
/// across all platforms.
/// </summary>
public enum ProbeType
{
    DataContainer,   // Table / DataGrid / List / Tree
    Selector,        // ComboBox / Select / Dropdown
    Action,          // Button / Link / MenuItem
    Display,         // Label / Badge / Counter / Chip
    Media,           // Image / Video / Audio / Canvas
    Form,            // Input / TextArea / Checkbox / Radio
    Page,            // Page-level container
    Modal,           // Dialog / Drawer / Popover / Tooltip
    Navigation,      // Tabs / Breadcrumb / Sidebar / Paginator
}

/// <summary>
/// Effect produced on a linkage target when the source element changes.
/// </summary>
public enum LinkageEffect
{
    OptionsUpdate,
    DataReload,
    VisibilityToggle,
    EnabledToggle,
    ValueUpdate,
    Reset,
    Navigate,
}

/// <summary>
/// Unified probe element model (15 properties + type-level extensions).
/// All platforms map their native UI trees to this structure.
/// Empty/null fields are omitted from serialization.
/// </summary>
public record ProbeElement
{
    // === Element Registry ===
    public required string Id { get; init; }
    public required ProbeType Type { get; init; }
    public AccessibilityInfo? Accessibility { get; init; }

    // === State Exposure ===
    public required StateInfo State { get; init; }

    // === Data (structure varies by Type) ===
    public DataInfo? Data { get; init; }

    // === Source Binding ===
    public SourceInfo? Source { get; init; }

    // === Linkage ===
    public LinkageInfo? Linkage { get; init; }

    // === Layout Metrics ===
    public required LayoutInfo Layout { get; init; }

    // === Shortcuts ===
    public IReadOnlyList<ShortcutInfo>? Shortcuts { get; init; }

    // === Animation ===
    public AnimationInfo? Animation { get; init; }

    // === Locale ===
    public LocaleInfo? Locale { get; init; }

    // === Theme ===
    public ThemeInfo? Theme { get; init; }

    // === Event Bindings ===
    public IReadOnlyList<string>? EventBindings { get; init; }

    // === Session ===
    public SessionInfo? Session { get; init; }

    // === Hierarchy ===
    public string? Parent { get; init; }
    public IReadOnlyList<string>? Children { get; init; }
}

public record AccessibilityInfo
{
    public string? Role { get; init; }
    public string? Label { get; init; }
    public int? TabIndex { get; init; }
}

public record StateInfo
{
    public required string Current { get; init; }
    public string? Previous { get; init; }
    public long Timestamp { get; init; }
    public bool? IsOpen { get; init; }
    public IReadOnlyList<ValidationError>? ValidationErrors { get; init; }
}

public record ValidationError
{
    public required string Field { get; init; }
    public required string Message { get; init; }
}

/// <summary>
/// Data payload — structure depends on the element's ProbeType.
/// DataContainer: rows, columns, sort, filter, selectedRows.
/// Media: currentTime, duration, readyState, paused, networkState.
/// General: value, options.
/// </summary>
public record DataInfo
{
    public object? Value { get; init; }
    public IReadOnlyList<string>? Options { get; init; }
    public int? Rows { get; init; }
    public IReadOnlyList<ColumnInfo>? Columns { get; init; }
    public SortInfo? Sort { get; init; }
    public IReadOnlyList<FilterInfo>? Filter { get; init; }
    public IReadOnlyList<string>? SelectedRows { get; init; }

    // Media extensions
    public double? CurrentTime { get; init; }
    public double? Duration { get; init; }
    public int? ReadyState { get; init; }
    public bool? Paused { get; init; }
    public int? NetworkState { get; init; }
}

public record ColumnInfo
{
    public required string Id { get; init; }
    public required string Label { get; init; }
    public bool Visible { get; init; }
}

public record SortInfo
{
    public required string Column { get; init; }
    public required string Direction { get; init; } // "asc" | "desc"
}

public record FilterInfo
{
    public required string Field { get; init; }
    public required string Operator { get; init; }
    public object? Value { get; init; }
}

public record SourceInfo
{
    public required string Url { get; init; }
    public required string Method { get; init; }
    public int? Status { get; init; }
    public double? ResponseTime { get; init; }
    public object? Payload { get; init; }
}

public record LinkageInfo
{
    public required IReadOnlyList<LinkageTarget> Targets { get; init; }
}

public record LinkageTarget
{
    public required string Id { get; init; }
    public required LinkageEffect Effect { get; init; }
    public required LinkagePath Path { get; init; }
}

/// <summary>
/// Base class for linkage path types. Discriminated by <see cref="PathType"/>.
/// </summary>
public abstract record LinkagePath
{
    public abstract string PathType { get; }
}

public record DirectPath : LinkagePath
{
    public override string PathType => "direct";
}

public record ApiPath : LinkagePath
{
    public override string PathType => "api";
    public required string Url { get; init; }
    public string? Method { get; init; }
}

public record ComputedPath : LinkagePath
{
    public override string PathType => "computed";
    public required string Expression { get; init; }
}

public record StorePath : LinkagePath
{
    public override string PathType => "store";
    public required string StoreName { get; init; }
    public string? Action { get; init; }
}

public record NavigationPath : LinkagePath
{
    public override string PathType => "navigation";
    public required string Route { get; init; }
}

public record ChainPath : LinkagePath
{
    public override string PathType => "chain";
    public required string Through { get; init; }
}

public record LayoutInfo
{
    public double X { get; init; }
    public double Y { get; init; }
    public double Width { get; init; }
    public double Height { get; init; }
    public bool Visible { get; init; }
    public double? RenderTime { get; init; }
    public double? ScrollTop { get; init; }
    public double? ScrollLeft { get; init; }
}

public record ShortcutInfo
{
    public required string Key { get; init; }
    public required string Action { get; init; }
    public string? Platform { get; init; }
}

public record AnimationInfo
{
    public bool Playing { get; init; }
    public string? Name { get; init; }
    public double? Duration { get; init; }
    public double? Progress { get; init; }
}

public record LocaleInfo
{
    public string? Language { get; init; }
    public string? Direction { get; init; } // "ltr" | "rtl"
    public bool? IsRTL { get; init; }
}

public record ThemeInfo
{
    public string? Mode { get; init; } // "light" | "dark" | "high-contrast"
    public string? ColorScheme { get; init; }
}

public record SessionInfo
{
    public bool? IsDirty { get; init; }
    public bool? HasUnsavedChanges { get; init; }
}

/// <summary>
/// Platform context describing the runtime environment.
/// </summary>
public record PlatformContext
{
    public required Platform Platform { get; init; }
    public required DeviceProfile Device { get; init; }
    public required ViewportSize Viewport { get; init; }
    public required InputMode InputMode { get; init; }
}

public record ViewportSize
{
    public int Width { get; init; }
    public int Height { get; init; }
}

public enum Platform
{
    WebChrome,
    WebSafari,
    WebFirefox,
    Ios,
    Android,
    MacOs,
    Windows,
    Linux,
}

public record DeviceProfile
{
    public required string Name { get; init; }
    public required ScreenSize ScreenSize { get; init; }
    public double PixelRatio { get; init; }
    public bool HasNotch { get; init; }
    public bool HasSafeArea { get; init; }
    public required string FormFactor { get; init; } // "phone" | "tablet" | "desktop" | "foldable"
}

public record ScreenSize
{
    public int Width { get; init; }
    public int Height { get; init; }
}

public enum InputMode
{
    Touch,
    MouseKeyboard,
    Stylus,
    Gamepad,
}
